Document `tsh` aliases usage. Change implementation details. (#14931)
gravitational/teleport: docs/pages/use-teleport/tsh.mdx

In either of those files you can add define an alias such as:

```yaml
aliases:
    "l": "tsh login --auth=okta --username=alice"
```

From now on, `tsh l` will resolve to `tsh login --auth=okta --username=bob`.

You can also change defaults for regular commands:

```yaml
aliases:
    "status": "tsh status --format=json"
```

Calling external programs other than `tsh` is also possible:

```yaml
aliases:
    "connect": "bash -c 'tsh login $0 && tsh ssh $1'"
```

The example above also demonstrates the usage of variables `$0` and `$1`. They represent arguments provided to the alias. With the definition above `tsh connect foo bar` resolves to `bash -c 'tsh login foo && tsh ssh bar'`. The alias can use as many arguments as needed. If the alias is invoked with too few arguments, `tsh` will catch that and report is as an error. Conversely providing additional arguments is *NOT* an error; any unused arguments will simply be appended at the end of the expanded alias definition.

The alias substitution happens before the command line flags are fully parsed. This means that it is not affected by the `--debug` flag. To debug aliases, set the `TELEPORT_DEBUG=1` environment variable instead.

> If you want to temporarily disable an alias, set the `TSH_ALIAS` environment variable. `tsh` understands the value of that flag to be a comma-separated list of aliases it should ignore. For example, `TSH_ALIAS=connect,status` will make it behave as if those two aliases were not set. This environment variable is also used to prevent infinite alias loops from occurring.

Reviewer: These seems like low level implementation details, I wouldn't include them in the docs, or maybe at least move to a separate "troubleshooting" section.

Author: Yeah, happy to drop this. This is probably too much for the average user and more for the hackish folks that will look up the source code anyway.

  Extensions:         permit-X11-forwarding
```

## Custom aliases and defaults

> You can define your own custom commands and set command-specific flag defaults by using the `aliases` section of `tsh` config.

Reviewer 2: ```suggestion
You can configure `tsh` to define aliases, custom commands and command-specific flag defaults. Using aliases, you can run frequently used `tsh` commands more easily.
```
We haven't introduced the `tsh` config file yet, so I think we should lead with the benefit of this feature.

> There are two configuration files:

Reviewer 2: ```suggestion
`tsh` loads two kinds of configuration files:
```

> - global: by default `/etc/tsh.yaml`, unless overridden by `$TELEPORT_GLOBAL_TSH_CONFIG` env var.

Reviewer 2: ```suggestion
- global: by default `/etc/tsh.yaml`, unless overridden by the `$TELEPORT_GLOBAL_TSH_CONFIG` env var.
```

- user-specific: `$TELEPORT_HOME/config/config.yaml`, which by defaults resolves to `~/.tsh/config/config.yaml`.

> The user-specific config is merged with global config. In case of conflicts (i.e. same alias defined in both files), the user-specific config has higher priority.

Reviewer 2: ```suggestion
`tsh` merges the user-specific config with the global config. In case of conflicts (i.e. same alias defined in both files), the user-specific config has higher priority.
```

> The user-specific config is merged with global config. In case of conflicts (i.e. same alias defined in both files), the user-specific config has higher priority.

Reviewer 2: I think we should move this paragraph lower in the section. Readers don't know what an alias configuration looks like at this point in the section, and I think this would be good to know before we talk about merging configs.

In either of those files you can add define an alias such as:

```yaml
aliases:
    "l": "tsh login --auth=okta --username=alice"
```

> 

Reviewer 2: As with the config reference, I think we should include more general information about how `tsh` handles the alias config. Does each key become a separate subcommand? Are there any restrictions on keys?

From now on, `tsh l` will resolve to `tsh login --auth=okta --username=bob`.

> You can also change defaults for regular commands:

Reviewer 2: ```suggestion
You can also change the defaults for regular `tsh` commands:
```

```yaml
aliases:
    "status": "tsh status --format=json"
```

Calling external programs other than `tsh` is also possible:

```yaml
aliases:
    "connect": "bash -c 'tsh login $0 && tsh ssh $1'"
```

> The example above also demonstrates the usage of variables `$0` and `$1`. They represent arguments provided to the alias. With the definition above `tsh connect foo bar` resolves to `bash -c 'tsh login foo && tsh ssh bar'`. The alias can use as many arguments as needed. If the alias is invoked with too few arguments, `tsh` will catch that and report is as an error. Conversely providing additional arguments is *NOT* an error; any unused arguments will simply be appended at the end of the expanded alias definition.

Reviewer 2: ```suggestion
The example above demonstrates the usage of variables `$0` and `$1`. They represent arguments provided to the alias. With the definition above, `tsh connect foo bar` resolves to `bash -c 'tsh login foo && tsh ssh bar'`. 

The alias can use as many arguments as needed. If the alias is invoked with too few arguments, `tsh` will report an error. Conversely, providing additional arguments is *not* an error. `tsh` will append any additional arguments to the end of an alias definition.
```

Reviewer 2: I think it would be useful to include a code snippet illustrating additional arguments that exceed the number of parameters in an alias.

> An additional variable that can be used in alias definition is `$TSH`. It will be expanded to absolute path to current `tsh` executable. This can be useful if there are multiple `tsh` versions installed, or the currently used version is not in `PATH`.

Reviewer 2: ```suggestion
You can also add the `$TSH` variable to an alias definition. When invoking the alias, `tsh` will expand this to the absolute path to current `tsh` executable. This can be useful if there are multiple `tsh` versions installed, or the currently used version is not in `PATH`.
```
